File: DS using c/stack-applications/postfix-evaluation.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
#define MAX_SIZE 100
double stack[MAX_SIZE];
int top = -1;
double pop()
{
    if (top != -1)
        return stack[top--];
    else
        return '#';
}
void push(double j)
{
    if(top==MAX_SIZE-1){
        printf("Stack is full");
        exit(0);
    }
    stack[++top] = j;
}
char peek()
{
    return stack[top];
}

int is_operator(char ch)
{
    return ch == '+' || ch == '*' || ch == '-' || ch == '^' || ch == '/';
}

int is_digit(char ch){
    return ch>='0' && ch<='9';
}

void postfix_evaluvation(char postfix[])
{
    int postfix_value = 0;
    for (int i=0;i<strlen(postfix);i++)
    {
        if (!is_operator(postfix[i]) && !is_digit(postfix[i]))
        {
            printf("Invalid Expression!!!");
            return;
        }
        else if (is_digit(postfix[i]))
        {
            push(postfix[i] - '0');
        }
        else if (is_operator(postfix[i]))
        {
            double operand2 = pop();
            double operand1 = pop();
            switch (postfix[i])
            {
            case '+':
                push(operand1 + operand2);
                break;
            case '*':
                push(operand1 * operand2);
                break;
            case '/':
                push(operand1 / operand2);
                break;
            case '-':
                push(operand1 - operand2);
                break;
            case '^':
                push(pow(operand1, operand2));
                break;
            }
        }
    }
    printf("postfix value is %.2f\n",(double)peek());
}

int main()
{
    char postfix[MAX_SIZE];
    printf("Enter postfix expression to evaluate: ");
    scanf("%s",postfix);
    postfix_evaluvation(postfix);
    return 0;
}

File: DS using python/stack-applications/postfix-evaluation.py
import array as arr


class Stack:
    def __init__(self):
        self.MAXSIZE = 100
        self.stack = arr.array("f", [0 for x in range(self.MAXSIZE)])
        self.top = -1
        self.postfix = ""

    def pop(self):
        if self.top != -1:
            popped = self.stack[self.top]
            self.top -= 1
            return popped
        else:
            return "#"

    def push(self, j):
        if self.top == self.MAXSIZE - 1:
            print("stack is full")
            exit(0)
        self.top += 1
        self.stack[self.top] = j

    def peek(self):
        return self.stack[self.top]

    def is_operator(self, ch):
        return ch == "+" or ch == "*" or ch == "-" or ch == "^" or ch == "/"

    def postfix_evaluation(self, postfix):
        for i in postfix:
            if not self.is_operator(i) and not i.isdigit():
                print("invalid expression")
            elif i.isdigit():
                self.push(ord(i) - 48)
            elif self.is_operator(i):
                operand2 = self.pop()
                operand1 = self.pop()
                if i == "+":
                    self.push(operand1 + operand2)
                elif i == "*":
                    self.push(operand1 * operand2)
                elif i == "/":
                    self.push(operand1 / operand2)
                elif i == "-":
                    self.push(operand1 - operand2)
                elif i == "^":
                    self.push(operand1 ** operand2)
        print("Postfix value is %.2f" % self.peek())


if __name__ == "__main__":
    postfix = input("Enter postfix expression : ")
    stack = Stack()
    stack.postfix_evaluation(postfix)
